File: go/goapp/json_test.go
package main

import (
	"encoding/json"
	"fmt"
	"testing"
)

// 常用 Tag 参数说明
// omitempty 将字段值为空值 "" 0 nil false 时，忽略该字段，不参与序列化
// omitzero  将字段值为零值 "" 0 nil false 与包括自定义类型和嵌套结构体的零值时，忽略该字段，不参与序列化（Go 1.24 新特性）
// string    将数值类型 int、float、bool 转换为 JSON 字符串类型，反序列化时自动反向转换
// inline    将嵌套结构体字段时，将嵌套结构体的字段“扁平化”到父结构体 JSON 中
// -         将忽略该字段，不参与序列化与反序列化

// 使用 omitzero 的场景
// 处理时间类型：time.Time 的零值处理是 omitzero 最典型的应用场景
// 区分 nil 和空集合：需要保留空数组 [] 但忽略 nil 时使用
// 自定义零值判断：通过 IsZero() 方法实现业务特定的零值逻辑
//
// 使用 omitempty 的场景
// 常规字符串字段：空字符串需要被忽略的场景
// 指针类型：nil 指针需要被忽略的场景
// 简单的布尔和数字字段：false 和 0 需要被忽略的场景
//
// 需要精确控制零值用 omitzero，常规空值忽略用 omitempty

type Query struct {
	Path   string  `json:"path,omitempty"`
	Query  string  `json:"query,omitempty"`
	Age    int     `json:"age,string,omitempty"` // 可以是 {"age":"18"} 但不能是  {"age":18}
	Score  float64 `json:"score,string,omitempty"`
	IsPass bool    `json:"is_pass,string"`
}

func TestJson(t *testing.T) {
	query := Query{
		Path:  "/index/index",
		Query: "name=laixhe&age=18",
		Age:   18,
		Score: 88.8,
	}
	jsonBytes, err := json.Marshal(query)
	fmt.Println(string(jsonBytes), err)
	// 结果 {"path":"/index/index","query":"name=laixhe\u0026age=18","age":"18","score":"88.8","is_pass":"false"}

	query2 := Query{}
	jsonBytes2 := []byte(`{"path":"/index/index","query":"name=laixhe&age=19","age":"19","score":"99.99","is_pass":"true"}`)
	err = json.Unmarshal(jsonBytes2, &query2)
	fmt.Println(query2, err)
}

File: java/javalog/src/test/java/com/laixhe/javalog/JsonTest.java
package com.laixhe.javalog;

import com.laixhe.javalog.demo.JsonDemo;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * JSON 序列化测试（对应 Go json_test.go 的核心断言）。
 */
class JsonTest {

    // 练习 1：omitempty —— 空值/空集合忽略（对应 Go TestJson）
    @Test
    void exercise1_omitempty() throws Exception {
        JsonDemo.TJson tJson = new JsonDemo.TJson(null, List.of(), Map.of());
        String json = JsonDemo.MAPPER.writeValueAsString(tJson);
        assertEquals("{}", json); // 全部被忽略

        JsonDemo.TJson full = new JsonDemo.TJson(
                java.time.Instant.parse("2025-06-21T09:18:39Z"), List.of(1, 2), Map.of("a", "1"));
        String jsonFull = JsonDemo.MAPPER.writeValueAsString(full);
        assertTrue(jsonFull.contains("\"time1\":\"2025-06-21T09:18:39Z\""));
        assertTrue(jsonFull.contains("\"array1\":[1,2]"));
    }

    // 练习 2：omitzero 变体 —— 只忽略 null，空集合保留
    @Test
    void exercise2_omitzero() throws Exception {
        JsonDemo.TJsonOmitZero omitZero = new JsonDemo.TJsonOmitZero(List.of(), Map.of());
        String json = JsonDemo.MAPPER.writeValueAsString(omitZero);
        assertTrue(json.contains("\"array2\":[]"));
        assertTrue(json.contains("\"map2\":{}"));
    }

    // 练习 3：string tag —— 数值以字符串序列化（对应 Go TestJsonQuery）
    @Test
    void exercise3_number_as_string() throws Exception {
        JsonDemo.Query query = new JsonDemo.Query("/index/index", "name=laixhe", 18, 88.8, false);
        String json = JsonDemo.MAPPER.writeValueAsString(query);
        assertTrue(json.contains("\"age\":\"18\""));
        assertTrue(json.contains("\"score\":\"88.8\""));
        assertTrue(json.contains("\"is_pass\":\"false\""));
    }

    // 练习 4：反序列化 —— 字符串数字自动转回数值类型
    @Test
    void exercise4_deserialize() throws Exception {
        String input = "{\"path\":\"/index/index\",\"query\":\"name=laixhe&age=19\","
                + "\"age\":\"19\",\"score\":\"99.99\",\"is_pass\":\"true\"}";
        JsonDemo.Query query = JsonDemo.MAPPER.readValue(input, JsonDemo.Query.class);
        assertEquals("/index/index", query.path());
        assertEquals(19, query.age());
        assertEquals(99.99, query.score());
        assertTrue(query.isPass());
    }

    // 练习 5：@JsonIgnore 忽略敏感字段（对应 Go "-" tag）
    @Test
    void exercise5_ignore() throws Exception {
        JsonDemo.Secret secret = new JsonDemo.Secret("laixhe", "123456");
        String json = JsonDemo.MAPPER.writeValueAsString(secret);
        assertEquals("{\"name\":\"laixhe\"}", json);
        assertFalse(json.contains("123456"));
    }

    // 运行完整 Demo
    @Test
    void runJsonDemo() throws Exception {
        JsonDemo.jsonBasic();
        JsonDemo.jsonQuery();
        JsonDemo.jsonPretty();
    }
}
